psi4/src/psi4/libmints/matrix.cc

```diff
@@ -3403,8 +3403,34 @@ SharedMatrix doublet(const SharedMatrix &A, const SharedMatrix &B, bool transA,
 
 SharedMatrix triplet(const SharedMatrix &A, const SharedMatrix &B, const SharedMatrix &C, bool transA, bool transB,
                      bool transC) {
-    SharedMatrix T = doublet(A, B, transA, transB);
-    SharedMatrix S = doublet(T, C, false, transC);
+
+    int dim1 = transA ? A->ncol() : A->nrow();
+    int dim2 = transA ? A->nrow() : A->ncol();
+    int dim3 = transB ? B->ncol() : B->nrow();
+    int dim4 = transB ? B->nrow() : B->ncol();
+    int dim5 = transC ? C->ncol() : C->nrow();
+    int dim6 = transC ? C->nrow() : C->ncol();
```

Comment:
I suspect `doublet` is calling the symmetry-aware 6-arg form of `gemm` (https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/matrix.cc#L1347), so I wonder if this is ok and if should be calling `max_ncol` etc. instead (https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/matrix.h#L640). better consult @jturney

Reply:
Good point: the optimized `triplet` should actually contain a loop over irreps and determine the evaluation sequence one irrep at a time, since the sizes of the submatrices may not match.

Reply:
@susilehtola Awesome! I will work on it tomorrow!

Reply:
I just made the changes :)

Reply:
No, that's still just a partial solution. What you need to do is compute the symmetry blocks separately once at a time, since the optimal order `(AB)C` vs `A(BC)` depends on the symmetry block.

Reply:
i.e. you have the correct cost for each `h`, now what you need is to call `CBLAS` to evaluate the blocks of the matrix transform in the right order.

Reply:
@susilehtola Ah, I talked to Zach and Lori about this earlier today. We agreed that this would be too much of a hassle. Although I'm not sure about CBLAS... @zachglick @loriab Thoughts?

Reply:
I don't think it'd be a hassle; the relevant part of `gemm()` is something like 6 lines of code.

Reply:
I think it'd need much of https://github.com/psi4/psi4/blob/87b0117c345f257ea763de92d0462e93e82e62d9/psi4/src/psi4/libmints/matrix.cc#L1359-L1378 over single irrep, and we'd want to avoid duplication, I'd guess.

`double` and `triplet` were Rob/Ed (Parrish/Hohenstein) concoctions so the use is heavily c1 codes of scf, sapt, dfmp2. but there are a few mcscf, dct calls will possible symmetry. bottom line, existing scheme is an improvement and is the maximum improvement for most present usages. but if `triplet` is to be more than easy syntax, getting the full efficiency out of it may be worth implementing.

Reply:
I'll take a stab at it, doesn't seem to be too bad

Reply:
maybe the `gemm` code can be refactored so that you make a routine that does the `gemm` only for a single irrep, and change the current `gemm` function to calling that in a `for` loop.

Reply:
I like the idea